scilpy/segment/voting_scheme.py

```diff
@@ -0,0 +1,435 @@
+                continue
+            else:
+                logging.info('%s final recognition got %s streamlines',
+                             bundle_names[bundle_id], len(streamlines_id))
+
+            streamlines = tractogram.streamlines[streamlines_id.T]
+            vote_score = streamlines_wise_vote[streamlines_id.T, bundle_id]
+
+            # All models of the same bundle have the same basename
+            basename = os.path.join(self.output_directory,
+                                    os.path.splitext(bundle_names[bundle_id])[0])
+            out_tractogram = nib.streamlines.Tractogram(streamlines,
+                                                        affine_to_rasmm=np.eye(4))
+            nib.streamlines.save(out_tractogram, basename + extension,
+                                 header=tractogram.header)
+
+            curr_results_dict = {}
+            curr_results_dict['indices'] = np.asarray(streamlines_id).tolist()
+            curr_results_dict['votes'] = np.squeeze(
+                vote_score.toarray()).tolist()
+            results_dict[basename] = curr_results_dict
+
+        out_logfile = os.path.join(self.output_directory, 'results.json')
+        with open(out_logfile, 'w') as outfile:
+            json.dump(results_dict, outfile)
+
+    def multi_recognize(self, input_tractogram_path, tractogram_clustering_thr,
+                        nb_points=20, nbr_processes=1, seeds=None):
+        """
+        Parameters
+        ----------
+        input_tractogram_path : str
+            Filepath of the whole brain tractogram to segment
+        tractogram_clustering_thr : int
+            Distance in mm (for QBx) to cluster the input tractogram
+        nb_points : str
+            Number of points used for all resampling of streamlines
+        nbr_processes : int
+            Number of processes used for the parallel bundle recognition
+        seeds : list
+            List of seed for the RandomState
+        """
+
+        # Load the subject tractogram
+        timer = time()
+        tractogram = nib.streamlines.load(input_tractogram_path)
+        wb_streamlines = tractogram.streamlines
+        logging.debug('Tractogram %s with %s streamlines ' +
+                      'is loaded in %s seconds', input_tractogram_path,
+                      len(tractogram.streamlines),
+                      round(time() - timer, 2))
+
+        # Prepare all tags to read the atlas properly
+        bundle_names, bundles_filepath = self._init_bundles_tag()
+
+        # Cluster the whole tractogram only once per possible clustering threshold
+        rbx_all = {}
+        base_thresholds = [45, 35, 25]
+        for seed in seeds:
+            rng = np.random.RandomState(seed)
+            for clustering_thr in tractogram_clustering_thr:
+                timer = time()
+                # If necessary, add an extra layer (more optimal)
+                if clustering_thr < 15:
```

Comment:
Why 15 ? I don't understand

Reply:
It is because QBx requires threshold that are decreasing, but also having a useless layer is suboptimal. Instead of systematically adding a new layer or having corner case, there is a check. 

Imagine you have a clustering threshold of **8**: [45, 35, 25, 8] (big gap, suboptimal)
But I can't systematically add a layer at 15 in case some use a clustering threshold of **18**: [45, 35, 25, 15, 18] (Inversion)

So basically the logic is just to generate layer by skip of ten, but the last one has a safety check because it is not always necessary